src/core/quantum_gate.jl

```diff
@@ -1464,6 +1754,38 @@ get_operator(gate::Toffoli, T::Type{<:Complex}=ComplexF64) = toffoli(T)
 
 get_connected_qubits(gate::Toffoli)=[gate.control_1, gate.control_2, gate.target]
 
+get_control_qubits(gate::Toffoli)=[gate.control_1, gate.control_2]
+
+get_target_qubits(gate::Toffoli)=[gate.target]
+
+# optimized application of ControlX, ControlZ or Toffoli gate without calling operator 
+# (it is hard-coded in apply_control_x!, apply_control_z! or apply_toffoli!, respectively)
+function apply_gate!(state::Ket, gate::Union{ControlX,ControlZ,Toffoli})
```

Comment:
This doesn't look great since this is essentially three functions and we doing a union to share code which looks like a code smell. What we are in fact repeating is the validation step which we have in every `apply_gate`.

We can refactor (but in a later PR) `apply_gate` into `apply_gate` into `unsafe_apply_gate` where `apply_gate` does the validation and then passes the values onto the `unsafe_apply_gate` function. All the shared validation then happens once. (All gates are required to be inside the circuit) and it dispatches to `unsafe_apply_gate` which can then handle the gate specifics. Then we don't have to have Union constructs and then by-hand dispatch to the correct function.

EDIT:

I've created an issue for the refactor so that we can do it later: #214.